### ISSUE TITLE
Personal note - updated note got appended after refresh

### ISSUE THREAD

**Issue by huhugrub** [cgeo/cgeo#6590]
##### Detailed steps causing the problem:
- Step 1 Edit Personal Cache Note and Save on the geocaching.com listing page.
- Step 2 Refresh listing in c:geo

##### Actual behavior after performing these steps:
Updated Personal Cache Note get appended to the Personal note in Cache Description
Old personal note remains in Cache Description

##### Expected behavior after performing these steps:
Old personal note should be replaced by updated personal cache note after refresh.

##### Version of c:geo used:
2017.05.22

##### Is the problem reproducible for you?
Yes

##### System information:
```
Attach system information here if available (see c:geo menu -> About c:geo -> Swipe right to System)
Keep the apostrophe at beginning and end to have it properly formatted
```

##### Other comments and remarks:
e.g. Reference to other issues, projects, sources, etc.

**Comment by pstorch** [comment]
The problem is to determine which Note is the most current. If the user forgets to upload and does a refresh, it would be fatal to replace his local Note with the older one from the server.

**Comment by huhugrub** [comment]
Perhaps one approach is on refresh, compare the server and local notes.
1.  If server is blank, upload the local note to server.
2.  If local is blank, download the server note to local.
3.  If server = local, then no action is needed.
4.  If server <> local, then ask user to select Keep Server, Keep Local, or Append Server to Local.

**Comment by pstorch** [comment]
@huhugrub it's already done like that, but sometimes the comparison fails. Maybe due to changes in whitespaces (line endings) on the server side. I think to fix it we would need to "normalize" the texts before comparing them.

**Comment by huhugrub** [comment]
@pstorch In the case I encountered, I believe the comparison had detected that the server note and the local note were different.  So what I got was the server note got appended to the local note in c:geo.  I understand that might be the intended behaviour "to play safe", if the refresh does not know which note is more current.  By step 4 above, I was suggesting perhaps in this case, to have a dialogue to ask the user to indicate which version to keep in c:geo (server or local).  I'm not sure if there is such a step in the current release.

**Comment by pstorch** [comment]
@huhugrub think of someone who refreshes a list of 1000 caches. How many "Personal-Note-Diff-Dialogs" do you want to show?

One possible solution I can think of is to introduce a new flag in the DB to indicate if the personal note is locally only or was already uploaded to the server. If it was already uploaded, we could safely overwrite the local one with a newer one from the server. Then it would be clear, I guess.

**Comment by Lineflyer** [comment]
Yes, in case they are different we keep both. No other chance to determine which is newer, and asking on each refresh is too much IMHO.

**Comment by huhugrub** [comment]
Different users might have different usage styles or preferences.  Some will primarily use the server as the master, while others primarily update their notes in the field using c:geo.  Perhaps, introduce a global preference setting to give users the choice.  E.g.
1.  If server=local, then no note update is necessary.
2.  If local is blank, then update local with server note.
3.  If server note is blank, then update server with local note.
4.  If server<>local, then proceed according to global note refresh setting.

Where global note refresh setting has following options:
A.  Always update local from server.
B.  Always update server from local.
C.  Always append server note to local note, where different.
D.  Always ask user, where different.

I understand Option C is the current behaviour.  But the problem with this is that multiple refresh updates will keep appending the server notes and the local notes will keep growing and cause confusion.  Guess that is up to the user to tidy this up manually, but it will mean a simple update in the server would still necessite a further step to manually edit out the old version in c:geo after the refresh.   But the above global settings will provide the flexibility in giving users the choice to suit their individual usage patterns.

As for bulk refreshes of many caches, the user can always choose Option C to continue with the current behaviour, if they don't want to be presented with multiple ask dialogues.

I agree Option D (to ask users which one to keep) will make it more complicated, but it is the only sure safe way to allow users to resolve the note version contention in those cases where it is needed.  But at least Options A and B will be useful features to cater for many users who primarily update their notes in either the server or the local instances according to their different usage patterns.

**Comment by WulfmanDo** [comment]
Hello,  what is to think about two Note Files in c:geo.

One with the Server Note ( none edditable, only show) , every time a refesh (Reload in c:geo) ist started, refresh the Notes in the Server notes field, so you can always see what have you saved at the Server in your notes.

In the second Notes Field, that are the (editable field) c:geo  notes, which you can edit in the fields.

And then the buttons, like, "edit", "save"  and "upload to server"  ( after a upload to server, the first note (server notes, will be refresh and be equal to the second note field)

So  never it can be happend that you kill with a refresh (or a refresh all caches in Folder)  your server or local notes. 

And when you get home after yor caching trip and got wlan/online you can  upload your c:geo notes ( create in the field)  to the server  by hand or by bluk  (select the caches you did today) > export > personal notes  ( like it is now embedded)

Greetings from Germany
Marco Wulfman_Do

**Comment by Lineflyer** [comment]
Having two seperate fields would save this conflict case indeed. Then it would only be a question of a appropriate UI, which is also not so easy?!

**Comment by pstorch** [comment]
I still think my idea with the a new flag could help here. Let's introduce a "personal_note_dirty" flag. As long as it is false the personal note from server overwrites the local one. As soon as the user updates it locally it becomes true. Now we perform the merge as now, because we don't know which one is newer. When the user uploads the personal note we reset this flag to false. Now the local changes are synchronized to the server and we can again safely overwrite the local version. 
This would reduce the number of append-merges to the time between local changes occurred until they are uploaded.
